src/layers/domain/core/device.py

```diff
@@ -0,0 +1,351 @@
+    ) -> DevicePage:
+        """
+        Create a page for storing values
+        """
+        if not re.match(PAGE_NAME_REGEX, index):
+            raise InvalidKeyError(f"Invalid page index: {index}")
+        if index in self.pages:
+            raise DuplicateError(f"Page exists: {index}")
+        page = DevicePage(values=values or {})
+        self.add_event(DevicePageAddedEvent(id=self.id, page=index))
+        self.pages[index] = page
+        return page
+
+    def remove_page(self, index: str = ""):
+        """
+        Remove a page from the device
+        """
+        if index not in self.pages:
+            raise NotFoundError(f"Unknown page: {index}")
+        self.add_event(DevicePageRemovedEvent(id=self.id, page=index))
+        del self.pages[index]
+
+    def get_page(self, index: str = "") -> DevicePage:
+        """
+        Retrieve a single page
+        """
+        if index not in self.pages:
+            raise NotFoundError(f"Unknown page: {index}")
+        return self.pages[index]
+
+    def set_values(self, values: dict[str, DevicePageValue], page: str = None):
+        """
+        Update the values on a page
+        """
+        page = self.get_page(page or "")
+        page.values = {**page.values, **values}
+        self.add_event(DevicePageUpdatedEvent(id=self.id, page=page))
+
+    def add_relationship(
+        self, target: "Device", type: RelationshipType
+    ) -> RelationshipAddedEvent:
+        """
+        Adds a relationship between two Devices
+        """
+        if target.id in self.relationships:
+            raise DuplicateError(f"Relationship {target.id} exists")
+        self.relationships[target.id] = Relationship(type=type)
+        event = RelationshipAddedEvent(id=self.id, target=target.id, type=type)
+        return self.add_event(event=event)
+
+    def remove_relationship(self, target: str) -> RelationshipRemovedEvent:
+        """
+        Removes an existing relationship between two Devices
+        TODO: Do we delete them or mark them inactive?
+        """
+        target_id = target.id if isinstance(target, Device) else target
+        if target_id not in self.relationships:
+            raise NotFoundError(f"Relationship {target_id} is unknown")
+        del self.relationships[target_id]
+        event = RelationshipRemovedEvent(id=self.id, target=target_id)
+        return self.add_event(event=event)
+
+    def add_key(self, key: ProductId | AccreditedSystemId) -> DeviceKeyAddedEvent:
+        """
+        Adds a new Key to a Device
+        """
+        if key in self.keys:
+            raise DuplicateError(f"Key {key} exists")
+        if isinstance(key, ProductId):
+            type = DeviceKeyType.PRODUCT_ID
+        elif isinstance(key, AccreditedSystemId):
+            type = DeviceKeyType.ACCREDITED_SYSTEM_ID
+        else:
+            raise InvalidDeviceKeyError(
+                "key must be ProductId() or AccreditedSystemId()"
+            )
```

Comment:
for extensibility i'd have each key know it's own `DeviceKeyType`

Reply:
The relationship is not bidirectional.  DVK relies on PID/ASID, but not the other way round.

Reply:
yes exactly - i'm suggesting key type can be a property of the dvk that way only keys with types can be used